### DISCUSSION TITLE
Load config only if plugin is installed and loaded

### DISCUSSION THREAD

**Discussion by dstadelm** [wbthomason/packer.nvim#812] [Q&A] [answered]
Thank you for this very nice package manager.

When installing fresh I get all sorts of errors and the plugins fail to install. The reason are configs that fail due to the missing plugins. I then comment out all 'config = ' lines and then I am able to install the plugins.
Why are the configs loaded at all when the plugin isn't yet installed?

**Comment by wbthomason** [comment, marked as answer]
This has been discussed before, and I implemented a tentative solution in https://github.com/wbthomason/packer.nvim/pull/508. However, we found that checking for an installed plugin before running its config led to flickering for some plugins (e.g. statuslines), and shelved the idea until we could come up with something better. If you have ideas for how to work around this problem, I'd be happy to hear them!

**Reply by dstadelm** [reply]
But in that case bootstraping is basically not possible, or am I missing something?
I'm not implying that this has to be possible, but the Readme led me to that conclusion. I also have some sequence issues during installation, but I think I can work around these.

What does "wants" do? I didn't find a description.

**Reply by dstadelm** [reply]
Also in your setup I can see that you use
```
config = function() require('...') end

or

config = [[ require('...')]]
```

Where I only do
```
config = require('...')
```

Am I missing something here?

**Reply by wbthomason** [reply]
Oh, I see now! Yes - you need to provide either a function or a string to the `config` key, which will be run at startup time from the compiled file. If you directly `require` your configuration as it seems you are, it will indeed be executed before your plugins are installed, breaking bootstrapping.

If you pass functions or strings, then on initial load run `:PackerSync` (or `packer.sync()`), your plugins will be installed, *then* the compiled file will be generated and sourced. On subsequent startups, your config will be loaded from the compiled file.

In the (hopefully near) future, the compiled file is going away, but you'll still need to provide a function/string to the `config` and `setup` keys to allow deferring execution to the appropriate time.

I hope this helps!

**Reply by dstadelm** [reply]
Thanks, that helps

**Reply by dstadelm** [reply]
Thank you for your help. I appreciate it very much. Currently im struggling with lspconfig and treesitter. The string I pass does not appear in the packer_compiled.lua and therefore do not work. I see in your dot files, that you are not configuring them by passing anything through config, but I can't figure out how you configure them...

**Reply by wbthomason** [reply]
Have you run `:PackerSync` or `:PackerCompile`?

I configure mine through files in the `plugin/` directory, which automatically get sourced by Neovim on startup.

**Comment by dstadelm** [comment]
I've just found out the reason. I added multiple packages without the require option, then config is not picked up
